services/topology-engine/queue-engine/topologylistener/messageclasses.py

```diff
@@ -39,85 +41,63 @@
-        try:
-            event_handled = False
-
-            if self.get_message_type() == MT_SWITCH:
-                if self.payload['state'] == "ADDED":
-                    event_handled = self.create_switch()
-                elif self.payload['state'] == "ACTIVATED":
-                    event_handled = self.activate_switch()
-                elif self.payload['state'] == "DEACTIVATED":
-                    event_handled = self.deactivate_switch()
-                elif self.payload['state'] == "REMOVED":
-                    event_handled = self.remove_switch()
-
-                if event_handled:
-                    message_utils.send_cache_message(self.payload,
-                                                     self.correlation_id)
-
-            elif self.get_message_type() == MT_ISL:
-                if self.payload['state'] == "DISCOVERED":
-                    event_handled = self.create_isl()
-                elif self.payload['state'] == "FAILED":
-                    event_handled = self.isl_discovery_failed()
-
-                if event_handled:
-                    message_utils.send_cache_message(self.payload,
-                                                     self.correlation_id)
-
-            elif self.get_message_type() == MT_PORT:
-                if self.payload['state'] == "DOWN":
-                    event_handled = self.port_down()
-                else:
-                    event_handled = True
-
-            elif self.get_message_type() == MT_FLOW_INFODATA:
-                event_handled = self.flow_operation()
-
-            elif self.get_command() == CD_NETWORK:
-                event_handled = self.dump_network()
-
-            if not event_handled:
-                logger.error('Message was not handled correctly: message=%s',
-                             self.payload)
-
-            return event_handled
-        except Exception as e:
-            logger.exception("Exception during handling message")
-            return False
+        elif self.kind == MT_ISL:
+            if self.payload['state'] == "DISCOVERED":
+                event_handled = self.create_isl()
+            elif self.payload['state'] == "FAILED":
+                event_handled = self.isl_discovery_failed()
+            else:
+                raise exc.NoHandlerError
+
+            if event_handled:
+                message_utils.send_cache_message(self.context, self.payload)
+
+        elif self.kind == MT_PORT:
+            if self.payload['state'] == "DOWN":
+                event_handled = self.port_down()
+            else:
+                event_handled = True
+
+        elif self.kind == MT_FLOW_INFODATA:
+            event_handled = self.flow_operation()
+
+        elif self.kind == CD_NETWORK:
+            event_handled = self.dump_network()
 
-#        finally:
-#            return True
+        else:
+            raise exc.NoHandlerError
+
+        if not event_handled:
+            raise exc.RecoverableError
```

Comment:
How do we know if it is Recoverable?

Reply:
We don't know. Code that throw this exception - know.

PS This retry mechanism is ugly and can bring more problem than solve issues. It should be refactored.